Microsoft.Quantum.Katas/KataMagic.cs

```diff
@@ -0,0 +1,168 @@
+                : ExecuteStatus.Error.ToExecutionResult();
+        }
+
+        /// <summary>
+        /// Compiles the given code. Checks there is only one operation defined in the code,
+        /// and returns its corresponding OperationInfo
+        /// </summary>
+        public virtual OperationInfo Compile(string code, IChannel channel)
+        {
+            try
+            {
+                var result = Snippets.Compile(code);
+
+                foreach (var m in result.warnings) { channel.Stdout(m); }
+
+                // Gets the names of all the operations found for this snippet
+                var opsNames =
+                    result.Elements?
+                        .Where(e => e.IsQsCallable)
+                        .Select(e => e.ToFullName().WithoutNamespace(Microsoft.Quantum.IQSharp.Snippets.SNIPPETS_NAMESPACE))
+                        .OrderBy(o => o)
+                        .ToArray();
+
+                if (opsNames.Length > 1)
+                {
+                    channel.Stdout("Expecting only one Q# operation in code. Using the first one");
+                }
+
+                return Resolver.Resolve(opsNames.First());
+            }
+            catch (CompilationErrorsException c)
+            {
+                foreach (var m in c.Errors) channel.Stderr(m);
+                return null;
+            }
+            catch (Exception e)
+            {
+                channel.Stderr(e.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Executes the given kata using the provided `userAnswer` as the actual answer.
+        /// To do this, it finds another operation with the same name but in the Kata's namespace
+        /// (by calling `FindRawAnswer`) and replace its implementation with the userAnswer
+        /// in the simulator.
+        /// </summary>
+        public virtual bool Simulate(OperationInfo kata, OperationInfo userAnswer, IChannel channel)
+        {
+            var rawAnswer = FindRawAnswer(kata, userAnswer);
+            if (rawAnswer == null) throw new InvalidOperationException($"Invalid task: {userAnswer.FullName}");
+
+            try
+            {
+                var qsim = CreateSimulator();
+
+                qsim.DisableLogToConsole();
+                qsim.Register(rawAnswer.RoslynType, userAnswer.RoslynType, typeof(ICallable));
+                qsim.OnLog += channel.Stdout;
+
+                var value = kata.RunAsync(qsim, null).Result;
+
+                if (qsim is IDisposable dis) { dis.Dispose(); }
+
+                return true;
+            }
+            catch (AggregateException agg)
+            {
+                foreach (var e in agg.InnerExceptions) { channel.Stderr(e.Message); }
+                channel.Stderr($"Try again!");
+                return false;
+            }
+            catch (Exception e)
+            {
+                channel.Stderr(e.Message);
```

Comment:
It may be worth taking an `ILogger` in the constructor and logging exceptions here so that we can get better debugging info (including the full stack trace) if something goes wrong.

Reply:
I've added the ILogger and use it for unexpected exceptions. Most errors, though, are expected, so not really creating logs for those.
While doing so, clean up the code to remove Exception when user errors are expected.